packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt

```diff
@@ -512,15 +510,146 @@ class AmplifyDataStorePlugin : FlutterPlugin, MethodCallHandler {
 
 
     @VisibleForTesting
-    fun deserializeNestedModels(serializedModelData: Map<String, Any>): Map<String, Any> {
-        return serializedModelData.mapValues {
-            if (it.value is Map<*, *>) {
-                SerializedModel.builder()
-                    .serializedData(deserializeNestedModels(it.value as HashMap<String, Any>))
-                    .modelSchema(null)
+    fun deserializeNestedModel(serializedModelData: Map<String, Any?>, modelSchema: ModelSchema): Map<String, Any?> {
+        val result = mutableMapOf<String, Any?>()
+
+        // iterate over schema fields to deserialize each field value
+        for ((key, field) in modelSchema.fields.entries) {
+            // ignore if serializedModelData doesn't contain value for the current field key
+            if (!serializedModelData.containsKey(key)) {
+                continue
+            }
+            val fieldSerializedData = serializedModelData[key]
+
+            // if the field serialized value is null
+            // assign null to this field in the deserialized map
+            if (fieldSerializedData == null) {
+                result[key] = fieldSerializedData
+                continue
+            }
+
+            if (field.isModel) {
+                // ignore field if the field doesn't have valid schema in ModelProvider
+                val fieldModelSchema = modelProvider.modelSchemas()[field.targetType] ?: continue
```

Comment:
When would this be the case and should we ignore it?

Reply:
This case should never happen theoretically, added this branch to ensure `fieldModelSchema` type is `ModelSchema` instead of `ModelSchema?`.
For clarity could change to checking `null` with throwing exception.